mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1167,6 +1197,223 @@ pub async fn p2sh_spending_tx<T: UtxoCommonOps>(coin: &T, input: P2SHSpendingTxI
+
+    // TODO validate premium amount. Might be a bit tricky in the case of dynamic miner fee
+    // TODO validate that output amounts are larger than dust
+
+    // Here, we have to use the exact lock time and premium amount from the preimage because maker
+    // can get different values (e.g. if MTP advances during preimage exchange/fee rate changes)
+    let expected_preimage = gen_dex_fee_spend_preimage(
+        coin,
+        gen_args,
+        LocktimeSetting::UseExact(actual_preimage_tx.lock_time),
+        CalcPremiumBy::UseExactAmount(actual_preimage_tx.outputs[1].value),
+    )
+    .await?;
+    let redeem_script =
+        swap_proto_v2_scripts::dex_fee_script(gen_args.time_lock, gen_args.secret_hash, &taker_pub, &maker_pub);
+    let sig_hash = signature_hash_to_sign(
+        &expected_preimage,
+        DEFAULT_SWAP_VOUT,
+        &redeem_script,
+        coin.as_ref().conf.signature_version,
+        coin.as_ref().conf.fork_id,
+    )?;
+
+    if !taker_pub
+        .verify(&sig_hash, &preimage.signature.clone().into())
+        .map_to_mm(|e| ValidateDexFeeSpendPreimageError::SignatureVerificationFailure(e.to_string()))?
+    {
+        return MmError::err(ValidateDexFeeSpendPreimageError::InvalidTakerSignature);
+    };
+    let expected_preimage_tx: UtxoTx = expected_preimage.into();
+    if expected_preimage_tx != actual_preimage_tx {
+        return MmError::err(ValidateDexFeeSpendPreimageError::InvalidPreimage);
+    }
+    Ok(())
+}
+
+pub async fn sign_and_broadcast_dex_fee_spend<T: UtxoCommonOps>(
+    coin: &T,
+    preimage: &TxPreimageWithSig,
+    gen_args: &GenDexFeeSpendArgs<'_>,
+    secret: &[u8],
+    htlc_keypair: &KeyPair,
+) -> TransactionResult {
+    let taker_pub = try_tx_s!(Public::from_slice(gen_args.taker_pub));
+
+    let mut dex_fee_tx: UtxoTx = try_tx_s!(deserialize(gen_args.dex_fee_tx));
+    dex_fee_tx.tx_hash_algo = coin.as_ref().tx_hash_algo;
+    drop_mutability!(dex_fee_tx);
+
+    let mut preimage_tx: UtxoTx = try_tx_s!(deserialize(preimage.preimage.as_slice()));
+    preimage_tx.tx_hash_algo = coin.as_ref().tx_hash_algo;
+    drop_mutability!(preimage_tx);
+
+    let secret_hash = dhash160(secret);
+    let redeem_script = swap_proto_v2_scripts::dex_fee_script(
+        gen_args.time_lock,
+        secret_hash.as_slice(),
+        &taker_pub,
+        htlc_keypair.public(),
+    );
+
+    let mut signer: TransactionInputSigner = preimage_tx.clone().into();
+    signer.inputs[0].amount = dex_fee_tx.outputs[0].value;
+    signer.consensus_branch_id = coin.as_ref().conf.consensus_branch_id;
+    drop_mutability!(signer);
+
+    let maker_signature = try_tx_s!(calc_and_sign_sighash(
+        &signer,
+        DEFAULT_SWAP_VOUT,
+        &redeem_script,
+        htlc_keypair,
+        coin.as_ref().conf.signature_version,
+        coin.as_ref().conf.fork_id
+    ));
+    let sig_hash_all_fork_id = 1 | coin.as_ref().conf.fork_id as u8;
```

Comment:
Are we 100% sure that `fork_id` in coins configuration is always `<= u8::MAX`? Otherwise `sig_hash_all_fork_id` may have unexpected/random values.

Reply:
There is no such validation and coins repo even has asset using 2 bytes value for `fork_id`: 
https://github.com/KomodoPlatform/coins/blob/05eb26c9fdecb6168f4d900fa2db7464ddd6b675/coins#L7229.
I am now wondering whether this LTFN coin works at all 🙂

Reply:
UPD: at least it used to work at the moment of addition: https://github.com/KomodoPlatform/coins/commit/676149e0645fc2e94f25f45f4cf37123e480072a.

Reply:
The problem of using `as` is sometimes it returns random values when overflow occurs which may result in unexpected behaviours and it's really hard to catch.

AFAIK `fork_id` is currently `u32`. If we update it to `u8`, it will make more sense to use it `fork_id as u32` which can't overflow so there will be no unexpected behaviour. And if a longer value than `u8` was given to `fork_id` in coins config, mm2 will validate and panic the process automatically on the initialization

Reply:
The trick with forementioned LTFN coin is it already uses `0x2A40` fork_id that doesn't fit into u8, and it somehow worked with `as u8` conversion, which is actually used for a long time in our signing code:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/ef896149506899fd76f983bb82d9b210699c7707/mm2src/coins/utxo_signer/src/sign_common.rs#L118

I will need some time to do research on this, as it looks strange.